### ISSUE TITLE
Kinked branch lines with similar colours cannot be followed by eye

### ISSUE THREAD

**Issue by mikebeaton** [ForkIssues/TrackerWin#183]
Hi - yet another one, sorry! I hope this input is helpful and useful - I'm only creating (several of!) these because Fork is a GREAT product! (And currently free - btw, seriously, I would have happily donated something already, if you had a donate link - I looked for one! :) )

I noticed that you have these kind of kinked branch lines (visible in the screenshot on [your own homepage](https://forkapp.io/)),

What has happened on my repository is that a kinked branch line has got pushed right next to another branch line and then two next door branches have also randomly ended up with very similar colours, which makes the branch structure almost impossible to read:

![fork](https://user-images.githubusercontent.com/11946605/53634477-2c310a00-3c12-11e9-8f30-2d73828f0e1f.jpg)

In the above, it looks exactly as if commit `10281b9` has `30235f3` as a parent, but it doesn't; there's a kink in the line, and it then passes down to `1a8ec76` and then to the shared root of `fc99e60` (as it should for a merge).

To actually be sure what was going on, I had to open gitk, which showed it like this:

![gitk](https://user-images.githubusercontent.com/11946605/53634476-2b987380-3c12-11e9-9468-905fee755103.jpg)

As I say, the fact that two next door branches randomly got very similar colours makes the branch lines very hard to follow by eye - perhaps some code to prevent that happening would be possible to add? I think that maybe arguably the use of the 'kinks' in the branch lines maybe doesn't help either? (But I can certainly see that there won't be any simple of perfect solution to something like this!) [EDIT: And/or, I wonder if there might possibly be a way to prevent kinks getting pushed too close to the near-horizontal parts of other branches, which perhaps would also have kept these two apart?)

I want to say, once again, many thanks for a lovely product!

**Comment by DanPristupov** [comment]
Hi

It it seems you are using topo-order in Fork and date-order in gtk. Could you try to switch that in Fork preferences and show the screenshot again?

**Comment by mikebeaton** [comment]
Hi Dan,

Actually it's in date order in Fork. I could try to revert (a local copy of) the repository back to this state and send an image in both sort orders, if it would help?

Mike

**Comment by mikebeaton** [comment]
(But surely one could generate _some_ potential commit tree with exactly the branch structure I've got, for either sort order?)

**Comment by DanPristupov** [comment]
Drawing graph is hard, however I'll be happy to improve it. I can dig into this one day, but I need to see same info, at least.

On your screenshots in Fork `HttpsTrue` is above `cors` and in gitk it's opposite. I can't give any precise answer seeing them.

I suppose this issue can be related to that one in the Mac tracker: https://github.com/ForkIssues/Tracker/issues/345

Update: it seems to be exactly the same issue.

**Comment by mikebeaton** [comment]
Okay - I see - no problem, of course! I'll get the same image in the same ordering on both. (I think I need to revert the repo to that state and then switch gitk to date ordering, but whatever it is, I'll look into it!)

**Comment by mikebeaton** [comment]
Hi, I've reset my repo to something similar (at least, showing the same issue), and forced gitk to display the commits in the same (date) order as Fork. (I've also added a screenshot of Bitbucket's version, again in the same order.)

Here is the Fork display. For me, it is very hard to follow the path from 0223a0e down to fc99e60 (it looks very much as if it joins back in at 30235f5, when it does not).

![screenshot_1](https://user-images.githubusercontent.com/11946605/53733135-ec6b5c00-3e77-11e9-8e7a-a4385759b538.jpg)

I think that this is for two reasons (i.e. that changing either would help): 1. Because the 'kink' has been allowed to occur nestled in right where the rest of the tree gets crowded, and 2. Because adjacent branches have been allocated very similar colours. (I do appreciate this stuff is very fiddly; that neither of these proposed 'fixes' might be at all trivial; that they might have other side effects I haven't thought of! And I appreciate your work in making this!!)

Here is the same tree in the same order on gitk:

![screenshot_2](https://user-images.githubusercontent.com/11946605/53733137-ef664c80-3e77-11e9-87e4-ecb87e4176d7.jpg)

And here in Bitbucket:

![screenshot_3](https://user-images.githubusercontent.com/11946605/53733141-f1301000-3e77-11e9-8206-722459c9d00f.jpg)

Hope these are more helpful in being directly like-for-like.

Thank you!

**Comment by DanPristupov** [comment]
OK, I made some improvements.

Fork 1.29
<img width="409" alt="image" src="https://user-images.githubusercontent.com/618115/54642753-b217d680-4a95-11e9-8487-e46800254d23.png">

Fork 1.30 (should be released by the end of the week) 
<img width="409" alt="image" src="https://user-images.githubusercontent.com/618115/54642811-c5c33d00-4a95-11e9-9cab-1f6e91b570f7.png">

**Comment by mikebeaton** [comment]
Many thanks indeed for all your the work on this.

The less sharp kinks do indeed make the branches easier to follow, so thanks!

If I can say, to me, looking at the improved page you've shown just above, the commit "Add iframe attributes to React" still does have hard-to-follow lines (which blue line goes in to join the node on orange, and which one carries on down?). If horizontally adjacent branches we forced to always have significantly different colours, I still think that might well additionally help.

Many thanks again. I still like Fork very much! Thank you.

**Comment by jerone** [comment]
[**@MikeBeaton**](https://github.com/MikeBeaton) commented on [Mar 20, 2019, 11:00 PM GMT+1](https://github.com/ForkIssues/TrackerWin/issues/183#issuecomment-475045480 "2019-03-20T22:00:47Z - Replied by Github Reply Comments"):
> If I can say, to me, looking at the improved page you've shown just above, the commit "Add iframe attributes to React" still does have hard-to-follow lines (which blue line goes in to join the node on orange, and which one carries on down?). If horizontally adjacent branches we forced to always have significantly different colours, I still think that might well additionally help.
> 
Agree. Maybe the color pallet could be extended with more colors.

> Many thanks again. I still like Fork very much! Thank you.

Also agree :)

**Comment by DanPristupov** [comment]
I just published the 1.30 update!

**Comment by sm-g** [comment]
please do not limit number of colors!

![Fork_2020-01-29_17-36-58](https://user-images.githubusercontent.com/7280273/73365965-37773480-42be-11ea-891b-ab6c8267f1d6.png)

this also may be related https://github.com/ForkIssues/TrackerWin/issues/308#issuecomment-500355335
